src/Processors/Transforms/Streaming/ShufflingTransform.cpp

```diff
@@ -124,29 +124,29 @@ void ShufflingTransform::consume(Chunk chunk)
 {
     if (chunk.hasRows())
     {
-        auto chunk_ctx = chunk.getOrCreateChunkContext();
+        auto chunk_ctx = chunk.getChunkContext(); /// save chunk context before split
         auto split_chunks{chunk_splitter.split(chunk)};
         for (auto & chunk_with_id : split_chunks)
         {
             assert(chunk_with_id.chunk);
             auto output_idx = chunk_with_id.id.items[0] % outputs.size();
             /// Keep substream id for each sub-chunk, used for downstream processors
-            auto new_chunk_ctx = std::make_shared<ChunkContext>(*chunk_ctx);
-            new_chunk_ctx->setSubstreamID(std::move(chunk_with_id.id));
-            chunk_with_id.chunk.setChunkContext(std::move(new_chunk_ctx));
+            chunk_with_id.chunk.setChunkContext(chunk_ctx ? chunk_ctx->clone() : nullptr);
+            chunk_with_id.chunk.getOrCreateChunkContext()->setSubstreamID(std::move(chunk_with_id.id));  
             shuffled_output_chunks[output_idx].push(std::move(chunk_with_id.chunk));
         }
     }
     else
     {
         chunk.trySetSubstreamID(INVALID_SUBSTREAM_ID);
 
-        /// Shuffling is very upstream, downstream substream watermark transform still
+        /// Shuffling is upstream, downstream substream watermark transform still
         /// depends on this empty timer chunk to calculate watermark for global aggregation
         /// When we fix the timer issue systematically, the pipeline system shall have minimum
         /// empty block flowing around and we don't need this anymore
+        auto chunk_ctx = chunk.getChunkContext();
         for (size_t output_idx = 0; output_idx < outputs.size(); ++output_idx)
-            shuffled_output_chunks[output_idx].push(chunk.clone());
+            shuffled_output_chunks[output_idx].push(chunk.cloneWithChunkContext(chunk_ctx));
```

Comment:
Do we need clone chunk context here ? I am thinking avoid cloning chunk context unless you need update  / modified it. We can established a new convention: if we need update chunk context, think about if we will need clone it since chunk context could be shared by other processors.

Reply:
Agree it, probablly we can use `COW shared ptr` for ChunkContext